File: src/auth/perms.rs
use std::collections::HashMap;

use crate::components::*;
use actix_web::web;
use log::*;
use regex::Regex;
use resalt_models::User;
use resalt_storage::StorageImpl;
use serde_json::Value;

pub const P_ADMIN_SUPERADMIN: &str = "admin.superadmin";
pub const P_ADMIN_GROUP: &str = "admin.group";

pub const P_RUN_LIVE: &str = "run.live";
#[allow(dead_code)]
pub const P_RUN_APPROVAL_LIST: &str = "run.approval.list";
#[allow(dead_code)]
pub const P_RUN_APPROVAL_SUBMIT: &str = "run.approval.submit";
#[allow(dead_code)]
pub const P_RUN_APPROVAL_JUDGE: &str = "run.approval.judge";
#[allow(dead_code)]
pub const P_RUN_APPROVAL_EXECUTE: &str = "run.approval.execute";
#[allow(dead_code)]
pub const P_RUN_APPROVAL_RECOUNT: &str = "run.approval.recount"; // Can reset all approvals/rejects
#[allow(dead_code)]
pub const P_RUN_APPROVAL_CLOSE: &str = "run.approval.close";
#[allow(dead_code)]
pub const P_RUN_APPROVAL_DELETE: &str = "run.approval.delete";
#[allow(dead_code)]
pub const P_RUN_TEMPLATE_LIST: &str = "run.template.list";
#[allow(dead_code)]
pub const P_RUN_TEMPLATE_LOCAL: &str = "run.template.local"; // Can create local templates
#[allow(dead_code)]
pub const P_RUN_TEMPLATE_GLOBAL: &str = "run.template.global"; // Can create global templates

pub const P_MINION_LIST: &str = "minion.list";
pub const P_MINION_CONFORMITY: &str = "minion.conformity";
pub const P_MINION_PILLARS: &str = "minion.pillars";
pub const P_MINION_PACKAGES: &str = "minion.packages";
#[allow(dead_code)]
pub const P_MINION_PRESETS_LIST: &str = "minion.presets.list";
#[allow(dead_code)]
pub const P_MINION_PRESETS_MANAGE: &str = "minion.presets.manage";
pub const P_MINION_GRAINEXPLORER: &str = "minion.grainexplorer";

pub const P_JOB_LIST: &str = "job.list";

pub const P_EVENT_LIST: &str = "event.list";

pub const P_SALTKEY_LIST: &str = "saltkey.list";
pub const P_SALTKEY_ACCEPT: &str = "saltkey.accept";
pub const P_SALTKEY_REJECT: &str = "saltkey.reject";
pub const P_SALTKEY_DELETE: &str = "saltkey.delete";

pub const P_USER_ADMIN: &str = "user";
pub const P_USER_LIST: &str = "user.list";
#[allow(dead_code)]
pub const P_USER_EMAIL: &str = "user.email";
pub const P_USER_PASSWORD: &str = "user.password";

pub fn has_resalt_permission(
    data: &web::Data<Box<dyn StorageImpl>>,
    user_id: &str,
    permission: &str,
) -> Result<bool, ApiError> {
    let user = match data.get_user_by_id(user_id) {
        Ok(user) => match user {
            Some(user) => user,
            None => return Ok(false),
        },
        Err(e) => {
            error!("{:?}", e);
            return Err(ApiError::DatabaseError);
        }
    };
    let perms = match serde_json::from_str(&user.perms) {
        Ok(perms) => perms,
        Err(e) => {
            error!("{:?}", e);
            return Err(ApiError::DatabaseError);
        }
    };
    Ok(evaluate_resalt_permission(&perms, permission))
}

pub fn evaluate_resalt_permission(permissions: &Value, permission: &str) -> bool {
    let args = Vec::new();
    let kwargs = HashMap::new();
    let normal = evaluate_permission(permissions, "@resalt", permission, &args, &kwargs);
    if !normal {
        evaluate_permission(permissions, "@resalt", P_ADMIN_SUPERADMIN, &args, &kwargs)
    } else {
        normal
    }
}

fn salt_wrapped_regex(regex: &str) -> String {
    format!("^{}$", regex.replace("([a-zA-Z0-9])\\*", "$1.*"))
}

fn evaluate_function(
    fun_section: &Value,
    fun: &str,
    args: &Vec<String>,
    kwargs: &HashMap<String, String>,
) -> bool {
    if let Some(fun_section) = fun_section.as_str() {
        let regex = salt_wrapped_regex(fun_section);
        let re = Regex::new(&regex).unwrap();
        return re.is_match(fun);
    }
    if let Some(fun_section) = fun_section.as_array() {
        if fun_section.len() == 0 {
            return true;
        }
        for arg in fun_section {
            let regex = salt_wrapped_regex(arg.as_str().unwrap());
            let re = Regex::new(&regex).unwrap();
            if re.is_match(&args.join(" ")) {
                return true;
            }
        }
        return false;
    }
    if let Some(fun_section) = fun_section.as_object() {
        if fun_section.contains_key("args") {
            let args_section = fun_section["args"].as_array().unwrap();
            if args_section.len() > 0 {
                for arg in args_section {
                    let regex = salt_wrapped_regex(arg.as_str().unwrap());
                    let re = Regex::new(&regex).unwrap();
                    if re.is_match(&args.join(" ")) {
                        return true;
                    }
                }
                return false;
            }
        }
        if fun_section.contains_key("kwargs") {
            let kwargs_section = fun_section["kwargs"].as_object().unwrap();
            if kwargs_section.len() > 0 {
                for key in kwargs_section.keys() {
                    let regex = salt_wrapped_regex(kwargs_section[key].as_str().unwrap());
                    let re = Regex::new(&regex).unwrap();
                    if re.is_match(&kwargs[key]) {
                        return true;
                    }
                }
                return false;
            }
        }
        return true;
    }
    false
}

fn evaluate_target(
    target_section: &Value,
    target: &str,
    fun: &str,
    args: &Vec<String>,
    kwargs: &HashMap<String, String>,
) -> bool {
    if let Some(target_section) = target_section.as_str() {
        let regex = salt_wrapped_regex(target_section);
        let re = Regex::new(&regex).unwrap();
        return re.is_match(fun);
    }
    let keys = target_section
        .as_object()
        .unwrap()
        .keys()
        .collect::<Vec<_>>();
    if keys.len() != 1 {
        return false;
    }
    for key in keys {
        let regex = salt_wrapped_regex(key);
        let re = Regex::new(&regex).unwrap();
        if re.is_match(target) {
            let fun_sections = target_section[key].as_array().unwrap();
            for fun_section in fun_sections {
                if evaluate_function(fun_section, fun, args, kwargs) {
                    return true;
                }
            }
            return false;
        }
    }
    false
}

pub fn evaluate_permission(
    permissions: &Value,
    target: &str,
    fun: &str,
    args: &Vec<String>,
    kwargs: &HashMap<String, String>,
) -> bool {
    let perms = match permissions.as_array() {
        Some(perms) => perms.to_vec(),
        None => Vec::new(),
    };
    for permission in perms {
        if evaluate_target(&permission, target, fun, &args, &kwargs) {
            return true;
        }
    }
    false
}

pub fn update_user_permissions_from_groups(
    data: &web::Data<Box<dyn StorageImpl>>,
    user: &User,
) -> Result<(), ApiError> {
    let groups = match data.list_permission_groups_by_user_id(&user.id) {
        Ok(groups) => groups,
        Err(e) => {
            error!("{:?}", e);
            return Err(ApiError::DatabaseError);
        }
    };
    let mut perms: Vec<Value> = Vec::new();
    for group in groups {
        // Parse group.perms as json array
        let serdegroup: serde_json::Value = match serde_json::from_str(&group.perms) {
            Ok(serdegroup) => serdegroup,
            Err(e) => {
                error!("{:?}", e);
                return Err(ApiError::DatabaseError);
            }
        };
        let group_perms = match serdegroup.as_array() {
            Some(group_perms) => group_perms,
            None => continue,
        };
        for group_perm in group_perms {
            perms.push(group_perm.clone());
        }
    }
    let perms = Value::Array(perms);
    let perms = serde_json::to_string(&perms).unwrap();
    let mut user: User = user.clone();
    user.perms = perms;
    match data.update_user(&user) {
        Ok(_) => Ok(()),
        Err(e) => {
            error!("{:?}", e);
            Err(ApiError::DatabaseError)
        }
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use serde_json::from_str;

    use crate::auth::evaluate_permission;
    use crate::auth::evaluate_resalt_permission;

    #[test]
    fn test_evalute_resalt_permission() {
        let perms = from_str(
            r#"[
                {
                  "minion\\*": [
                    "network.*"
                  ]
                },
                {
                  "@resalt": [
                    "admin.user.changepassword",
                    "admin.user.delete"
                  ]
                }
              ]"#,
        )
        .unwrap();
        assert!(!evaluate_resalt_permission(&perms, "test.ping"));
        assert!(evaluate_resalt_permission(
            &perms,
            "admin.user.changepassword"
        ));
        assert!(evaluate_resalt_permission(&perms, "admin.user.delete"));
        assert!(!evaluate_resalt_permission(&perms, "admin.unicorn"));
    }

    #[test]
    fn test_evalute_permission() {
        let perms = from_str(
            r#"[
                "test.ping",
                {
                  "*": [
                    {
                      "my_mod.*": {
                        "args": [
                          "a.*",
                          "b.*"
                        ],
                        "kwargs": {
                          "kwa": "kwa.*",
                          "kwb": "kwb"
                        }
                      }
                    }
                  ]
                },
                {
                  "@runner": [
                    {
                      "runner_mod.*": {
                        "args": [
                          "a.*",
                          "b.*"
                        ],
                        "kwargs": {
                          "kwa": "kwa.*",
                          "kwb": "kwb"
                        }
                      }
                    }
                  ]
                },
                {
                  "@wheel": [
                    "key.list_all"
                  ]
                },
                {
                  "G@os:RedHat": [
                    "kmod.*"
                  ]
                },
                {
                  "minion2": [
                    "network.*",
                    "state.*"
                  ]
                },
                {
                  "minion\\*": [
                    "network.*"
                  ]
                },
                {
                  "@resalt": [
                    "admin.user.changepassword",
                    "admin.user.delete"
                  ]
                }
              ]"#,
        )
        .unwrap();
        assert!(evaluate_permission(
            &perms,
            "minion1",
            "test.ping",
            &vec![],
            &HashMap::new()
        ));
        assert!(evaluate_permission(
            &perms,
            "minion1",
            "my_mod.my_fun",
            &vec!["a".to_string(), "b".to_string()],
            &HashMap::new()
        ));
        assert!(evaluate_permission(
            &perms,
            "minion1",
            "my_mod.my_fun",
            &vec!["a".to_string(), "b".to_string()],
            &vec![("kwa".to_string(), "kwa".to_string())]
                .into_iter()
                .collect()
        ));
        assert!(evaluate_permission(
            &perms,
            "minion1",
            "my_mod.my_fun",
            &vec!["a".to_string(), "b".to_string()],
            &vec![
                ("kwa".to_string(), "kwa".to_string()),
                ("kwb".to_string(), "kwb".to_string())
            ]
            .into_iter()
            .collect()
        ));
        assert!(!evaluate_permission(
            &perms,
            "minion1",
            "my_mod.my_fun",
            &vec!["a".to_string(), "b".to_string()],
            &vec![
                ("kwa".to_string(), "kwa".to_string()),
                ("kwb".to_string(), "kwc".to_string())
            ]
            .into_iter()
            .collect()
        ));
        assert!(!evaluate_permission(
            &perms,
            "minion1",
            "my_mod.my_fun",
            &vec!["a".to_string(), "b".to_string()],
            &vec![
                ("kwa".to_string(), "kwc".to_string()),
                ("kwb".to_string(), "kwb".to_string())
            ]
            .into_iter()
            .collect()
        ));
    }
}

File: frontend/src/lib/perms.ts
import type User from '../models/User';

export const P_ADMIN_SUPERADMIN = 'admin.superadmin';
export const P_ADMIN_GROUP = 'admin.group';

export const P_RUN_LIVE = 'run.live';
export const P_RUN_APPROVAL_LIST = 'run.approval.list';
export const P_RUN_APPROVAL_SUBMIT = 'run.approval.submit';
export const P_RUN_APPROVAL_JUDGE = 'run.approval.judge';
export const P_RUN_APPROVAL_EXECUTE = 'run.approval.execute';
export const P_RUN_APPROVAL_RECOUNT = 'run.approval.recount'; // Can reset all approvals/rejects
export const P_RUN_APPROVAL_CLOSE = 'run.approval.close';
export const P_RUN_APPROVAL_DELETE = 'run.approval.delete';
export const P_RUN_TEMPLATE_LIST = 'run.template.list';
export const P_RUN_TEMPLATE_LOCAL = 'run.template.local'; // Can create local templates
export const P_RUN_TEMPLATE_GLOBAL = 'run.template.global'; // Can create global templates

export const P_MINION_LIST = 'minion.list';
export const P_MINION_CONFORMITY = 'minion.conformity';
export const P_MINION_PILLARS = 'minion.pillars';
export const P_MINION_PACKAGES = 'minion.packages';
export const P_MINION_REFRESH = 'minion.refresh';
export const P_MINION_PRESETS_LIST = 'minion.presets.list';
export const P_MINION_PRESETS_MANAGE = 'minion.presets.manage';
export const P_MINION_GRAINEXPLORER = 'minion.grainexplorer';

export const P_JOB_LIST = 'job.list';
export const P_JOB_INFO = 'job.info';

export const P_EVENT_LIST = 'event.list';

export const P_SALTKEY_LIST = 'saltkey.list';
export const P_SALTKEY_ACCEPT = 'saltkey.accept';
export const P_SALTKEY_REJECT = 'saltkey.reject';
export const P_SALTKEY_DELETE = 'saltkey.delete';

export const P_USER_ADMIN = 'user';
export const P_USER_LIST = 'user.list';
export const P_USER_EMAIL = 'user.email';
export const P_USER_PASSWORD = 'user.password';

export const resaltWebPermissions: {
	permission: string;
	title: string;
	description: string;
	danger?: boolean;
	warning?: boolean;
}[] = [
	{
		permission: P_ADMIN_SUPERADMIN,
		title: '[Super Admin]',
		description: 'Grants ALL permissions in Resalt.',
		danger: true,
	},
	{
		permission: P_ADMIN_GROUP,
		title: '[Admin] Manage Groups',
		description: 'Allow to create, edit, and delete Groups in Resalt.',
		danger: true,
	},
	{
		permission: P_RUN_LIVE,
		title: '[Run] Live',
		description: 'Allow user to run live commands on minions they have access to.',
		danger: true,
	},
	{
		permission: P_RUN_APPROVAL_LIST,
		title: '[Run] [Approval] List',
		description: 'Allow user to list commands pending approval requests.',
	},
	{
		permission: P_RUN_APPROVAL_SUBMIT,
		title: '[Run] [Approval] Submit',
		description: 'Allow user to submit commands for approval.',
	},
	{
		permission: P_RUN_APPROVAL_JUDGE,
		title: '[Run] [Approval] Judge',
		description: 'Allow user to approve or deny approvals.',
		warning: true,
	},
	{
		permission: P_RUN_APPROVAL_EXECUTE,
		title: '[Run] [Approval] Execute',
		description: 'Allow user to execute approved approvals.',
		danger: true,
	},
	{
		permission: P_RUN_APPROVAL_RECOUNT,
		title: '[Run] [Approval] Recount',
		description: 'Allow user to reset approvals/rejects on a pending approval.',
		warning: true,
	},
	{
		permission: P_RUN_APPROVAL_CLOSE,
		title: '[Run] [Approval] Close',
		description: 'Allow user to close (and re-open) voting on approval request.',
	},
	{
		permission: P_RUN_APPROVAL_DELETE,
		title: '[Run] [Approval] Delete',
		description: 'Allow user to delete approval requests.',
		warning: true,
	},
	{
		permission: P_RUN_TEMPLATE_LIST,
		title: '[Run] [Templates] List',
		description: 'Allow user to list templates.',
	},
	{
		permission: P_RUN_TEMPLATE_LOCAL,
		title: '[Run] [Templates] Create Local',
		description: 'Allow user to create local templates.',
	},
	{
		permission: P_RUN_TEMPLATE_GLOBAL,
		title: '[Run] [Templates] Create Global',
		description: 'Allow user to create global templates.',
		warning: true,
	},
	{
		permission: P_MINION_LIST,
		title: '[Minion] List Minions',
		description: 'Allow user to list minions.',
	},
	{
		permission: P_MINION_CONFORMITY,
		title: '[Minion] See Conformity',
		description:
			'Allow user to see detailed conformity of minions. User can always see if minion is compliant or not.',
	},
	{
		permission: P_MINION_PILLARS,
		title: '[Minion] See Pillars',
		description: 'Allow user to see the pillars of minions.',
	},
	{
		permission: P_MINION_PACKAGES,
		title: '[Minion] See Packages',
		description: 'Allow user to see the packages of minions.',
	},
	{
		permission: P_MINION_REFRESH,
		title: '[Minion] Refresh',
		description:
			'Allow user to refresh minions. This is dependent on having minion permissions as well.',
		warning: true,
	},
	{
		permission: P_MINION_PRESETS_LIST,
		title: '[Minion] [Presets] List',
		description: 'Allow user to list minion presets.',
	},
	{
		permission: P_MINION_PRESETS_MANAGE,
		title: '[Minion] [Presets] Manage',
		description: 'Allow user to create, edit, and delete minion presets.',
		warning: true,
	},
	{
		permission: P_MINION_GRAINEXPLORER,
		title: '[Minion] Grain Explorer',
		description: 'Allow user to use the grain explorer.',
	},
	{
		permission: P_JOB_LIST,
		title: '[Job] List Jobs',
		description: 'Allow user to list jobs.',
	},
	{
		permission: P_EVENT_LIST,
		title: '[Event] List Events',
		description: 'Allow user to list events.',
	},
	{
		permission: P_SALTKEY_LIST,
		title: '[SaltKey] List All',
		description: 'Allow user to list salt keys.',
	},
	{
		permission: P_SALTKEY_ACCEPT,
		title: '[SaltKey] Accept',
		description: 'Allow user to accept salt keys.',
		danger: true,
	},
	{
		permission: P_SALTKEY_REJECT,
		title: '[SaltKey] Reject',
		description: 'Allow user to reject salt keys.',
		warning: true,
	},
	{
		permission: P_SALTKEY_DELETE,
		title: '[SaltKey] Delete',
		description: 'Allow user to delete salt keys.',
		danger: true,
	},
	{
		permission: P_USER_ADMIN,
		title: '[User] Admin',
		description: 'Allow user to create, manage, and delete other users.',
		danger: true,
	},
	{
		permission: P_USER_LIST,
		title: '[User] List All',
		description: 'Allow user to list users.',
	},
	{
		permission: P_USER_EMAIL,
		title: '[User] Email',
		description:
			'Allow user to change their own email. (LDAP users cannot change their email as it is synced via LDAP)',
	},
	{
		permission: P_USER_PASSWORD,
		title: '[User] Password',
		description:
			'Allow user to change their own password. (LDAP users cannot set or log in with local password)',
	},
];

function saltWrappedRegex(regex: string): RegExp {
	return new RegExp(`^${regex}$`.replace(/([a-zA-Z0-9])\*/g, '$1.*'));
}

type func = string;
type fKwargs = {
	[fun: string]: string;
};
type funcSectionArgs = {
	args: string[];
};
type funcSectionKwargs = {
	kwargs: fKwargs;
};
type funcSectionArgsKwargs = {
	args: string[];
	kwargs: fKwargs;
};
type funcSection =
	| func
	| {
			[fun: string]: string[] | funcSectionArgs | funcSectionKwargs | funcSectionArgsKwargs;
	  };
// "funSection" covers:
// - "fun"
// - { "fun": [] }
// - { "fun": { "args": [] } }
// - { "fun": { "kwargs": {} } }
// - { "fun": { "args": [], "kwargs": {} } }
type targetSection =
	| func
	| {
			[host: string]: funcSection[];
	  };

export function hasPermission(
	user: User | null,
	target: string,
	fun: string,
	args: string[],
	kwargs: fKwargs,
): boolean {
	if (!user) return false;
	// https://docs.saltproject.io/en/latest/topics/eauth/access_control.html
	// [
	//   {
	//     "@resalt": [
	//       "minion.list"
	//     ]
	//   },
	//   {
	//     "01*": [
	//       "grains.items"
	//     ]
	//   },
	//   "test.ping",
	//   {
	//     "bb": [
	//       "grains.items",
	//       "pkg.list"
	//     ]
	//   },
	//   {
	//     "cc": [
	//       {
	//         "grains.items": {
	//           "args": [
	//             "test"
	//           ],
	//           "kwargs": {
	//             "sanitize": true
	//           }
	//         }
	//       }
	//     ]
	//   },
	//   ".*",
	//   {
	//     ".*": [ //host
	//       "grains.items", //function
	//     ]
	//   },
	// ]
	// or
	// [
	//   "test.version",
	//   {
	// 	"mongo\\*": [
	// 	  "network.*"
	// 	]
	//   },
	//   {
	// 	"log\\*": [
	// 	  "network.*",
	// 	  "pkg.*"
	// 	]
	//   },
	//   {
	// 	"G@os:RedHat": [
	// 	  "kmod.*"
	// 	 ]
	//   }
	// ]

	let permissions: targetSection[] = [];
	if (user && user.perms) {
		permissions = user.perms as unknown as targetSection[];
	}

	// Both target and fun are REGEX, e.g "log*" or "pkg.*".

	const evaluateFunction = (
		funSection: funcSection,
		fun: string,
		args: string[],
		kwargs: fKwargs,
	): boolean => {
		if (typeof funSection === 'string') {
			const regex = saltWrappedRegex(funSection);
			return regex.test(fun);
		}
		const keys = Object.keys(funSection);
		if (keys.length !== 1) {
			return false;
		}
		for (const key of keys) {
			const regex = saltWrappedRegex(key);
			if (regex.test(fun)) {
				const value = funSection[key];
				if (typeof value === 'string') {
					return true;
				}
				if (Array.isArray(value)) {
					if (value.length === 0) {
						if (args.length !== 0) {
							return false;
						}
					}
					// Test each arg in the permission argainst "args"
					let result = true;
					for (let i = 0; i < value.length; i++) {
						const regex = saltWrappedRegex(value[i]);
						if (!regex.test(args[i])) {
							result = false;
							break;
						}
					}
					return result;
				}
				if (typeof value === 'object') {
					if ('args' in value) {
						if (value['args'].length === 0) {
							if (args.length !== 0) {
								return false;
							}
						}
						// Test each arg in the permission argainst "args"
						for (let i = 0; i < value['args'].length; i++) {
							const regex = saltWrappedRegex(value['args'][i]);
							if (!regex.test(args[i])) {
								return false;
							}
						}
					}
					if ('kwargs' in value) {
						const vvalue = value as funcSectionKwargs;
						const keys = Object.keys(vvalue.kwargs);
						if (keys.length === 0) {
							if (Object.keys(kwargs).length !== 0) {
								return false;
							}
						}
						// Test each arg in the permission argainst "kwargs"
						for (const key of keys) {
							const regex = saltWrappedRegex(vvalue.kwargs[key]);
							if (!regex.test(kwargs[key])) {
								return false;
							}
						}
					}
					return true;
				}
			}
		}
		return false;
	};

	const evaluateTarget = (
		targetSection: targetSection,
		target: string,
		fun: string,
		args: string[],
		kwargs: fKwargs,
	): boolean => {
		if (typeof targetSection === 'string') {
			const regex = saltWrappedRegex(targetSection);
			return regex.test(fun);
		}
		const keys = Object.keys(targetSection);
		if (keys.length !== 1) {
			return false;
		}
		for (const key of keys) {
			const regex = saltWrappedRegex(key);
			if (regex.test(target)) {
				const funSections = targetSection[key];
				for (const funSection of funSections) {
					if (evaluateFunction(funSection, fun, args, kwargs)) {
						return true;
					}
				}
				return false;
			}
		}
		return false;
	};

	for (const permission of permissions) {
		if (evaluateTarget(permission, target, fun, args, kwargs)) {
			return true;
		}
	}
	return false;
}

/**
 * Check if a user has a Resalt permission.
 * @param user The user to validate against, e.g. `$currentUser`.
 * @param permission The permission to check for, e.g. `P_USER_ADMIN`.
 * @returns True if the user has the permission, false otherwise.
 * @example
 * ```ts
 * if (hasResaltPermission($currentUser, P_USER_ADMIN)) {
 *  // Do something
 * }
 * ```
 */
export function hasResaltPermission(user: User | null, permission: string): boolean {
	return hasPermission(user, '@resalt', permission, [], {});
}

export default {};
